src/FSharp.Core/prim-types.fs

```diff
@@ -1654,27 +1654,149 @@ namespace Microsoft.FSharp.Core
+
+            let structuralEqualityComparer<'T> comparer =
+                { new EqualityComparer<'T>() with
+                    member _.Equals (x,y) =
+                        match box x, box y with 
+                        | null, null -> true
+                        | null, _    -> false
+                        | _,    null -> false
+                        | (:? IStructuralEquatable as x1), yobj -> x1.Equals (yobj, comparer)
+                        | _ -> raise (Exception "invalid logic - expected IStructuralEquatable")
+
+                    member _.GetHashCode x =
+                        match box x with 
+                        | null -> 0 
+                        | :? IStructuralEquatable as a -> a.GetHashCode fsEqualityComparerUnlimitedHashingPER
+                        | _ -> raise (Exception "invalid logic - expected IStructuralEquatable") }
+
+            let getEqualityComparer<'T> er =
+                let ty = typeof<'T>
+
+                match er with
+                | _ when canUseDefaultEqualityComparer er ty -> EqualityComparer<'T>.Default
+
+                | _ when isArray ty && er -> arrayEqualityComparer true fsEqualityComparerUnlimitedHashingER
+                | _ when isArray ty -> arrayEqualityComparer false fsEqualityComparerUnlimitedHashingPER
+                | _ when isStructuralEquatable ty && er -> structuralEqualityComparer fsEqualityComparerUnlimitedHashingER
+                | _ when isStructuralEquatable ty -> structuralEqualityComparer fsEqualityComparerUnlimitedHashingPER
+
+                | false when Type.op_Equality(ty, typeof<float>) -> unboxPrim (box {
+                    new EqualityComparer<float>() with
+                        member _.Equals (x,y) = (# "ceq" x y : bool #)
+                        member _.GetHashCode x = x.GetHashCode () })
+
+                | false when Type.op_Equality(ty, typeof<float32>) -> unboxPrim (box {
+                    new EqualityComparer<float32>() with
+                        member _.Equals (x,y) = (# "ceq" x y : bool #)
+                        member _.GetHashCode x = x.GetHashCode () })
+
+                | false -> { 
+                    new EqualityComparer<'T>() with
+                        member _.Equals (x,y) = GenericEqualityObj false fsEqualityComparerUnlimitedHashingPER (box x, box y)
+                        member _.GetHashCode x = GenericHashParamObj fsEqualityComparerUnlimitedHashingPER (box x) }
+
+                | true -> { 
+                    new EqualityComparer<'T>() with
+                        member _.Equals (x,y) = GenericEqualityObj true fsEqualityComparerUnlimitedHashingER (box x, box y)
+                        member _.GetHashCode x = GenericHashParamObj fsEqualityComparerUnlimitedHashingER (box x) }
+
+            type EqualityComparer<'T> () =
+
+                static member ERComparer = getEqualityComparer<'T> true
```

Comment:
This is a property which recomputes on every invocation - you want to use `static member val` or `let comparer ...` etc.

Reply:
Yep, that's fixed now.